src/main/java/com/aws/greengrass/deployment/ThingGroupHelper.java

```diff
@@ -0,0 +1,89 @@
+                    .retryableExceptions(Arrays.asList(SdkClientException.class)).build();
+
+    @Inject
+    public ThingGroupHelper(GreengrassServiceClientFactory clientFactory, DeviceConfiguration deviceConfiguration) {
+        this.clientFactory = clientFactory;
+        this.deviceConfiguration = deviceConfiguration;
+    }
+
+    /**
+     * Retrieve the thing group names the device belongs to.
+     * @return list of thing group names
+     * @throws InterruptedException if the thread is interrupted when fetching thing group list
+     * @throws NonRetryableDeploymentTaskFailureException  when not able to fetch thing group names
+     */
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException"})
+    public Optional<Set<String>> listThingGroupsForDevice()
+            throws InterruptedException, NonRetryableDeploymentTaskFailureException {
+
+        if (!deviceConfiguration.isDeviceConfiguredToTalkToCloud()) {
+            return Optional.empty();
+        }
+        try {
+            AtomicReference<String> nextToken = new AtomicReference<>();
+            Set<String> thingGroupNames = new HashSet<>();
+            return RetryUtils.runWithRetry(clientExceptionRetryConfig,
+                    () -> {
+                        do {
+                            ListThingGroupsForCoreDeviceRequest request = ListThingGroupsForCoreDeviceRequest.builder()
+                                    .coreDeviceThingName(Coerce.toString(deviceConfiguration.getThingName()))
+                                    .nextToken(nextToken.get()).build();
+                            ListThingGroupsForCoreDeviceResponse response = clientFactory.getGreengrassV2DataClient()
+                                    .listThingGroupsForCoreDevice(request);
+                            response.thingGroups().forEach(thingGroup -> {
+                                //adding direct thing groups
+                                thingGroupNames.add(THING_GROUP_RESOURCE_TYPE_PREFIX + thingGroup.thingGroupName());
+                                //adding parent thing groups
+                                thingGroup.rootToParentThingGroups().forEach(parentGroup -> thingGroupNames
+                                                .add(THING_GROUP_RESOURCE_TYPE_PREFIX + parentGroup.thingGroupName()));
+                            });
+                            nextToken.set(response.nextToken());
+                        } while (nextToken.get() != null);
+
+                        return Optional.of(thingGroupNames);
+                    },
+                    "get-thing-group-hierarchy", logger);
+        } catch (InterruptedException e) {
+            throw e;
+        } catch (Exception e) {
+            logger.atError().cause(e).log();
+            throw new NonRetryableDeploymentTaskFailureException("Error fetching thing group information", e);
```

Comment:
so why is this both logged and retrown?

Reply:
Logged so that we know the actual error. It is thrown so that we fail the deployment and the deployment failure reason is fetched from the exception message.

Reply:
That's really poor, that should be fixed, not hacked around

Reply:
My bad, the exception is logged in Deployment service. Removed the redundant log